src/vs/editor/contrib/copyPaste/browser/copyPasteController.ts

```diff
@@ -163,16 +162,28 @@ export class CopyPasteController extends Disposable implements IEditorContributi
 				dataTransfer.delete(vscodeClipboardMime);
 
 				for (const provider of [...providers, defaultPasteEditProvider]) {
-					const edit = await provider.provideDocumentPasteEdits(model, selection, dataTransfer, tokenSource.token);
+					if (!provider.pasteMimeTypes.some(type => {
+						if (type.toLowerCase() === DataTransfers.FILES.toLowerCase()) {
+							return [...dataTransfer.values()].some(item => item.asFile());
+						}
+						return dataTransfer.has(type);
+					})) {
+						continue;
+					}
+
+					const edit = await provider.provideDocumentPasteEdits(model, selections, dataTransfer, tokenSource.token);
 					if (originalDocVersion !== model.getVersionId()) {
 						return;
 					}
 
 					if (edit) {
-						if ((edit as WorkspaceEdit).edits) {
-							await this._bulkEditService.apply(ResourceEdit.convert(edit as WorkspaceEdit), { editor });
-						} else {
-							performSnippetEdit(editor, edit as SnippetTextEdit);
+						// TODO: batch this
+						for (const selection of selections) {
```

Comment:
@jrieken I am trying to apply a snippet to all cursor locations here. What's the best way to do this? I know that suggestions handle this but the code I found for it seems pretty complicated: https://github.com/microsoft/vscode/blob/b25b6dade8b67aa2209091803dce52210688ddc3/src/vs/editor/contrib/snippet/browser/snippetSession.ts#L437

Reply:
You can change the `performSnippetEdit`-function to take an array of ranges and simply set all of them as editor selection. The `constroller.insert(snippet)`-call will then take care of everything and do the right thing 🤞